sklearn/cluster/tests/test_optics.py

```diff
@@ -75,6 +78,40 @@ def test_close_extract():
     assert_equal(max(clust3.extract_dbscan(.3)[1]), 2)
 
 
+def test_dbscan_optics_parity():
+    # Test that OPTICS clustering metrics are < 2% difference of DBSCAN
+
+    centers = [[1, 1], [-1, -1], [1, -1]]
+    X, labels_true = make_blobs(n_samples=750, centers=centers,
+                                cluster_std=0.4, random_state=0)
+
+    # calculate optics with dbscan extract at 0.3 epsilon
+    op = OPTICS(min_samples=10).fit(X)
+    core_optics, labels_optics = op.extract_dbscan(0.3)
+
+    op_h = metrics.homogeneity_score(labels_true, labels_optics)
+    op_c = metrics.completeness_score(labels_true, labels_optics)
+    op_V = metrics.v_measure_score(labels_true, labels_optics)
+    op_m = metrics.adjusted_mutual_info_score(labels_true, labels_optics)
+    op_s = metrics.silhouette_score(X, labels_optics)
+
+    op_metrics = np.array([op_h, op_c, op_V, op_m, op_s])
+
+    # calculate dbscan labels and cluster metrics
+    db = DBSCAN(eps=0.3, min_samples=10).fit(X)
+
+    db_h = metrics.homogeneity_score(labels_true, db.labels_)
```

Comment:
Wouldn't we be better off comparing `db.labels_` and `op.labels_` directly, rather than testing that they are equally dissimilar from the ground truth? Probably best to just use one highly interpretable metric, or directly check a `contingency_matrix`. For `contingency_matrix`, we'd want the number of nonzero cells in each column or row to be almost zero (or the entropy in each row/column to be almost zero).

I'd also like to see in particular that they generate a near-identical set of outlier points.

Reply:
Yes, I definitely agree with you. The new unit test checks that OPTICS and DBSCAN agree on all core points, and that the mismatch for outlier point class labels is < 5 points (the test data is 750 points, ~10% of which are classified as either noise or outlier).

Reply:
Thanks!